metric_learn/constraints.py

```diff
@@ -33,6 +35,56 @@ def positive_negative_pairs(self, num_constraints, same_length=False,
       return a[:n], b[:n], c[:n], d[:n]
     return a, b, c, d
 
+  def generate_knntriplets(self, X, k_genuine, k_impostor):
```

Comment:
In the future we might want to rethink a bit the Constraints class, because this now requires access to `X`, not just the labels given at init. But let's keep things like this for now

Reply:
Agreed. It would be nicer to be able to pass `X` to `Constrains()`, maybe as an option for cases like this when it's needed.